python/feathub/dsl/ast.py

```diff
@@ -29,17 +64,48 @@ def to_json(self) -> Dict:
         """
         pass
 
+    @abstractmethod
+    def eval_dtype(self, variable_types: Dict[str, DType]) -> DType:
+        pass
+
     def __str__(self) -> str:
         return json.dumps(self.to_json(), indent=2, sort_keys=True)
 
 
-class BinaryOp(ExprAST):
-    def __init__(self, op_type: str, left_child: ExprAST, right_child: ExprAST) -> None:
-        super().__init__(node_type="BinaryOp")
-        self.op_type = op_type
+class AbstractUnaryOp(ExprAST, ABC):
+    def __init__(self, node_type: str, child: ExprAST):
+        super().__init__(node_type)
+        self.child = child
+
+
+class AbstractBinaryOp(ExprAST, ABC):
+    def __init__(self, node_type: str, left_child: ExprAST, right_child: ExprAST):
+        super().__init__(node_type)
         self.left_child = left_child
         self.right_child = right_child
 
+
+class BinaryOp(AbstractBinaryOp):
+    def __init__(self, op_type: str, left_child: ExprAST, right_child: ExprAST) -> None:
+        super().__init__(
+            node_type="BinaryOp", left_child=left_child, right_child=right_child
+        )
+        self.op_type = op_type
+
+    def eval_dtype(self, variable_types: Dict[str, DType]) -> DType:
+        left_type = self.left_child.eval_dtype(variable_types)
+        right_type = self.right_child.eval_dtype(variable_types)
+        if left_type == right_type:
+            return left_type
+
+        res_type = _get_higher_precision_type(left_type, right_type)
+        if res_type is None:
```

Comment:
I think by the signature of `eval_dtype`, it should return a non-None value.

However, even when dtypes of left and right children are non-None values, the `res_type` can be None. _get_higher_precision_type can return None if it sees a type other than Int and Float. For example, it can encounter this Feathub expr `'abc' * 'de'`

Reply:
Would it be simpler to move this check to `_get_higher_precision_type()`?